cmd/forwarder/auth.go

```diff
@@ -0,0 +1,212 @@
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	log "github.com/sirupsen/logrus"
+)
+
+// credentials are used by basic auth and include the hash of a valid password, plus
+// a "stage" string which is used to emit metrics that are useful when managing credrolls, so that
+// we can track whether or not deprecated passwords are still in use.
+type credential struct {
+	Stage string `json:stage`
```

Comment:
There aren't valid struct tags: https://play.golang.org/p/nELuf2qHQU7

Reply:
@freeformz interesting. i wonder how it's working? ... i'll investigate tomorrow. 

seriously though, i think it *is* working.

Reply:
It's working, just using `Stage` and `Hmac` as the keys vs. `stage` and `hmac`

Reply:
TIL `go vet`.

Reply:
Also, `json.Unmarshal` will match public members with appropriately named json hash keys even if there aren't any struct tags.